robot/web/stream/server.go

```diff
@@ -2,65 +2,52 @@ package webstream
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"sync"
 
 	"github.com/pion/webrtc/v3"
+	"github.com/pkg/errors"
 	"go.uber.org/multierr"
 	streampb "go.viam.com/api/stream/v1"
 	"go.viam.com/utils"
 	"go.viam.com/utils/rpc"
 
 	"go.viam.com/rdk/gostream"
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/robot/web/stream/state"
+	rutils "go.viam.com/rdk/utils"
 )
 
 type peerState struct {
-	stream  *streamState
-	senders []*webrtc.RTPSender
-}
-
-type streamState struct {
-	mu          sync.Mutex
-	stream      gostream.Stream
-	activePeers int
-}
-
-func (ss *streamState) Start() {
-	ss.mu.Lock()
-	defer ss.mu.Unlock()
-	ss.activePeers++
-	if ss.activePeers == 1 {
-		ss.stream.Start()
-	}
-}
-
-func (ss *streamState) Stop() {
-	ss.mu.Lock()
-	defer ss.mu.Unlock()
-	ss.activePeers--
-	if ss.activePeers <= 0 {
-		ss.activePeers = 0
-		ss.stream.Stop()
-	}
+	streamState *state.StreamState
+	senders     []*webrtc.RTPSender
 }
 
 // Server implements the gRPC audio/video streaming service.
 type Server struct {
 	streampb.UnimplementedStreamServiceServer
+	logger logging.Logger
+	r      robot.Robot
 
 	mu                      sync.RWMutex
-	streams                 []*streamState
-	nameToStream            map[string]gostream.Stream
+	streamNames             []string
+	nameToStreamState       map[string]*state.StreamState
 	activePeerStreams       map[*webrtc.PeerConnection]map[string]*peerState
 	activeBackgroundWorkers sync.WaitGroup
 }
 
 // NewServer returns a server that will run on the given port and initially starts with the given
 // stream.
-func NewServer(streams ...gostream.Stream) (*Server, error) {
+func NewServer(
+	streams []gostream.Stream,
+	r robot.Robot,
+	logger logging.Logger,
+) (*Server, error) {
 	ss := &Server{
-		nameToStream:      map[string]gostream.Stream{},
```

Comment:
There was no reason to have both a `map[string]gostream.Stream{}` and a `[]*streamState` so I consolidted them into `nameToStreamState`